iroh-bytes/src/store/bao_file.rs

```
@@ -0,0 +1,1238 @@
+        let res = Self {
+            data: bytes.to_vec().into(),
+            outboard: outboard.into(),
+            sizes: SizeInfo::complete(bytes.len() as u64),
+        };
+        (res, hash)
+    }
+
+    /// Persist the batch to disk, creating a FileBatch.
+    fn persist(&self, paths: DataPaths) -> io::Result<FileStorage> {
+        let mut data = create_read_write(&paths.data)?;
+        let mut outboard = create_read_write(&paths.outboard)?;
+        let mut sizes = create_read_write(&paths.sizes)?;
+        self.data.persist(&mut data)?;
+        self.outboard.persist(&mut outboard)?;
+        self.sizes.persist(&mut sizes)?;
+        data.sync_all()?;
+        outboard.sync_all()?;
+        sizes.sync_all()?;
+        Ok(FileStorage {
+            data,
+            outboard,
+            sizes,
+        })
+    }
+
+    pub(super) fn current_size(&self) -> u64 {
+        self.sizes.current_size()
+    }
+
+    pub(super) fn read_data_at(&self, offset: u64, len: usize) -> Bytes {
+        copy_limited_slice(&self.data, offset, len)
+    }
+
+    pub(super) fn data_len(&self) -> u64 {
+        self.data.len() as u64
+    }
+
+    pub(super) fn read_outboard_at(&self, offset: u64, len: usize) -> Bytes {
+        copy_limited_slice(&self.outboard, offset, len)
+    }
+
+    pub(super) fn outboard_len(&self) -> u64 {
+        self.outboard.len() as u64
+    }
+
+    pub(super) fn write_batch(
+        &mut self,
+        size: u64,
+        batch: &[BaoContentItem],
+    ) -> std::io::Result<()> {
+        let tree = BaoTree::new(ByteNum(size), IROH_BLOCK_SIZE);
+        for item in batch {
+            match item {
+                BaoContentItem::Parent(parent) => {
+                    if let Some(offset) = tree.pre_order_offset(parent.node) {
+                        let o0 = offset
+                            .checked_mul(64)
+                            .expect("u64 overflow multiplying to hash pair offset");
```

why panic, instead of returning an error?

---

this error would never happen. Even for a tree with data size u64::max, the offset will be something like data_size / chunk_group_size, so way below u64::max.

Maybe just using checked_mul here is overly paranoid and it should just be * 64.